nixos/modules/services/desktops/cpupower-gui.nix

```diff
@@ -0,0 +1,23 @@
+{ config, lib, pkgs, ... }:
+
+with lib;
+
+let
+  cfg = config.services.cpupower-gui;
+in {
+  options = {
+    services.cpupower-gui = {
+      enable = mkEnableOption "cpupower-gui";
+      description = ''
+        Enables dbus/systemd service needed used by cpupower-gui
+        to retrieve and modify cpu power saving settings.
+      '';
+    };
+  };
+
+  config = mkIf cfg.enable {
+    environment.systemPackages = [ config.boot.kernelPackages.cpupower-gui ];
+    services.dbus.packages = [ config.boot.kernelPackages.cpupower-gui ];
+    systemd.packages = [ config.boot.kernelPackages.cpupower-gui ];
```

Comment:
I don't know if this is right, I think we should convert the systemd unit files into their respective services:
```
cat ./results//linuxPackages.cpupower-gui/share/systemd/user/cpupower-gui-user.service
[Unit]
Description=Apply cpupower-gui config at user login

[Service]
Type=oneshot
ExecStart=/usr/bin/cpupower-gui --apply-config

[Install]
WantedBy=graphical-session.target
```
```
$ cat ./results/linuxPackages.cpupower-gui/lib/systemd/system/cpupower-gui-helper.service
[Unit]
Description=cpupower-gui system helper

[Service]
Type=dbus
BusName=org.rnd2.cpupower_gui.helper
ExecStart=/nix/store/babjq2zphc87j70qc97mqvb234lqhldx-cpupower-gui-0.9.1/lib/cpupower-gui/cpupower-gui-helper
```
```
$ cat ./results/linuxPackages.cpupower-gui/lib/systemd/user/cpupower-gui-user.service
[Unit]
Description=Apply cpupower-gui config at user login

[Service]
Type=oneshot
ExecStart=/usr/bin/cpupower-gui --apply-config

[Install]
WantedBy=graphical-session.target
```

Reply:
Oops. The first one definitely looks wrong. That `ExecStart` clearly won't work. Will fix.
The second is working here and gets triggered when the application is executed.

By converting, do you mean to explicitly define them as [part of the derivation](https://nixos.wiki/wiki/NixOS:extend_NixOS)?
My understanding is that if they are installed in the systemd folder, they will be picked up by systemd when installed. This is how the `dbus` systemd service is being executed. Though I agree that for the first, given it's a `oneshot` service, we could have a flag to enable it.

Reply:
What is the problem with the first service? This a user service and it can be enabled with `systemctl --user enable cpupower-gui-user.service`. 

It is `oneshot` as when you apply the settings, you don't need to re-apply. Unless something changed the settings, but then you would have to competing things trying to change the settings.

For the second service, it is used for automatic dbus activation. 
D-bus will start the helper when someone tries to connect to them.
By having the `SystemdService` key in `org.rnd2.cpupower_gui.helper.service` `dbus` will use systemd to manage the binary.
```
[D-BUS Service]
Name=org.rnd2.cpupower_gui.helper
Exec=@pkglibdir@/cpupower-gui-helper
User=root
SystemdService=cpupower-gui-helper.service
```

Reply:
@vagnum08 in NixOS there is no `/usr/bin/cpupower-bin` so the `ExecStart` path needs to modified to point to the correct location.

Reply:
Does it become something like `/nix/store/babjq2zphc87j70qc97mqvb234lqhldx-cpupower-gui-0.9.1/bin`.

If I change the service upstream to 
```
[Unit]
Description=Apply cpupower-gui config at user login

[Service]
Type=oneshot
ExecStart=@pkgbindir@/cpupower-gui --apply-config

[Install]
WantedBy=graphical-session.target
```
where `@pkgbindir@` is the path `prefix`/`bindir`  using meson's prefix/bindir variables. Will it work?

Reply:
nixos modules is able to define systemd units as nix code.

User unit example: https://github.com/NixOS/nixpkgs/blob/c0e65c63401c9c75c403df207e4bd6439b965cb4/nixos/modules/config/pulseaudio.nix#L280

System unit example: https://github.com/NixOS/nixpkgs/blob/c0e65c63401c9c75c403df207e4bd6439b965cb4/nixos/modules/config/pulseaudio.nix#L307

Reply:
The systemd unit files should be converted over to using nix, which is nix-aware.

The systemd unit files should then be cleaned up from the package.

 It's also common to expose a "package" option, which can be used to have a different version of the executed program. Example: https://github.com/NixOS/nixpkgs/blob/c0e65c63401c9c75c403df207e4bd6439b965cb4/nixos/modules/config/pulseaudio.nix#L145

Reply:
Thanks for the explanation.